paddle/fluid/imperative/layer.h

```diff
@@ -406,12 +420,54 @@ class OpBase : public std::enable_shared_from_this<OpBase> {
 
  private:
   OpBase(size_t id, const std::string& type, const NameVarBaseMap& ins,
-         const NameVarBaseMap& outs, framework::AttributeMap attrs,
+         const NameVarBaseMap& outs, const framework::AttributeMap& attrs,
          const platform::Place& place);
 
   OpBase(size_t id, const framework::OpDesc& op_desc,
          const platform::Place& place);
 
+ public:
+  OpBase() {}
+
+  void SetType(const std::string& type) { type_ = type; }
+  void SetInput(const std::string& name,
+                std::vector<std::shared_ptr<VarBase>> vec_var_base) {
+    ins_.emplace(name, vec_var_base);
```

Comment:
Do not use `emplace`. `emplace` would not insert or replace anything if the key has existed in the map. Use `insert` instead.

Reply:
first. Gradmaker should set the same name ONLY once. emplace have a batter performance than insert

Reply:
`ins_[name] = std::move(vec_var_base)`